src/main/java/net/neoforged/neoforge/common/util/CollectionReference.java

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.common.util;
+
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * An {@link AtomicReference} of {@link Collection} implementing {@link Collection}, providing a modifiable view of the
+ * referenced {@link Collection}.
+ * 
+ * @param <E> the element type of the {@link Collection}
+ * @param <C> the type of the {@link Collection}
+ */
+public class CollectionReference<E, C extends Collection<E>> extends AtomicReference<C> implements Collection<E> {
```

Comment:
This class is highly suspect.
Why do you need a AtomicReference to a collection? But then don't synchronize on the actual operation of modifying the collection?

Reply:
> I don't think using AtomicReferences is a great idea. It is much better to use concurrent collections, or synchronize on the field.

There's no need for concurrency or atomicity here, the field and the collection is not expecting any async operation anyways as a part of `MobEffectInstance`, so concurrent collections or `synchronize` is unrelated.
If `AtomicReference` seems too confusing, I'll switch to `MutableObject`.
The whole point of all this to provide a collection backed by a mutable delegate, more specifically, to keep the `MobEffectInstance#cures` as a final field.
If making the field non-final is acceptable (which breaks some original assumptions of the cures returned by getter, as the returned set may get outdated), I'll simply use the original `IdentityHashSet` anyways.